### ISSUE TITLE
Does PHP-DI have a feature where it automatically autowires an interface to one existing implementation?

### ISSUE THREAD

**Issue by robiningelbrecht** [PHP-DI/PHP-DI#867]
Been using PHP-DI for a while now and I really love it, simple, straight forward and fast.

One thing I keep stumbling upon though is that I need to map every interface to an implementation manually. I know that Symfony DI has a feature where it automatically "maps" an implementation to an interface if there is only one implementation of said interface. Is this something I can do with PHP-DI as well, or is this not supported?

**Comment by SvenRtbg** [comment]
That's quite hard to do automatically.

Not discussing the cached state where one could assume the result was magically baked into the config state... The only way to find which class implements an interface is to scan all existing classes, identify their interface implementation list, and then scan if the interface in question appears only once.

"All classes" may be different on various runtime environments, i.e. depending on extensions installed, it might not be a single class.

I don't know too much about Symfony DI, but I suspect they only consider their own packages.

In summary, while it is a bit annoying, the drawbacks are huge in several dimensions.

My personal recommendation would be to use fewer interfaces. YMMV.

**Comment by robiningelbrecht** [comment]
`My personal recommendation would be to use fewer interfaces. YMMV`

Not sure if joking or not :)
But ok, got my answer, thanks

**Comment by mnapoli** [comment]
> My personal recommendation would be to use fewer interfaces. YMMV

It's actually a good one. I've seen codebases (including my own in the past) where interfaces were created by default. But interfaces are often useless when there's only 1 implementation (i.e. ignoring libraries for example, but focusing mostly on company projects). It's often a result of "decoupling theatre", where it feels like having the interface brings any decoupling, but it actually doesn't do anything.

Anyway, that's how I interpreted this suggestion, maybe I missed the point :)

**Comment by SvenRtbg** [comment]
Your interpretation is on point. Having an interface with only one implementation, and basically no chance of there ever being two independent implementations, makes having an interface a useless habit.

**Comment by robiningelbrecht** [comment]
Making a lot of assumptions here :)

**Comment by SvenRtbg** [comment]
I have tried to find the documentation of Symfony where it explains the autowiring of interfaces, and I am convinced Symfony cannot do it. I found https://symfony.com/doc/current/service_container/autowiring.html#working-with-interfaces - and it states you have to configure the alias for the interface to be used - i.e. manually assign the implementation to be used to the interface.

In the next chapter, it is detailing how to have multiple implementations injected on a case-by-case basis - still needs manual definitions, and even more detailed and potentially brittle ones, using the parameter name as an additional hint. I believe PHP-DI might be unable to support this, unless you create a custom factory code (that allows you to do anything anyways).

I respect your code base, I have no knowledge about it. I made a generic statement about interfaces, and on the basis of having seen code that had interfaces everywhere, that I consider useless, and your inquiry about being able to automatically inject them, I obviously considered that this might be the result of too many interfaces that serve no real purpose. Hence, the "YMMV - your mileage may vary - your case may be entirely different and not subject to my comment at all". Sorry if I suggested anything different.

**Comment by robiningelbrecht** [comment]
No offense taken. Symfony can do it though, docs say:

`When using a service definition prototype, if only one service is discovered that implements an interface, configuring the alias is not mandatory and Symfony will automatically create one.`

I'm using this feature every day. I was just wondering if PHP-DI could do the same, nothing more, nothing less :)

**Comment by SvenRtbg** [comment]
I see. I missed that box during my first attempt - on the other hand, I assume these "service definition prototypes" still need registering, so there is manual effort involved to say to Symfony: Hey, please look at these selected set of classes - maybe they are useful when trying to autowire something.

And this selection of classes is something that will cost effort/performance if it requires too much code scanning, and on the other hand gives Symfony a head start, because it limits the potential candidates from "I don't even know where the files are that may contain code" to "I know exactly the file location and code I have to scan".

PHP-DI has no mechanism to allow for such an optimized lookup. I suspect the entire mechanism is highly connected to Symfony, targeting specific use cases inside this framework. But there is no such thing as "specialized, categorized classes" in a generic code base that PHP-DI is addressing. The documentation title reads "PHP-DI: The dependency injection container for humans", and I translate this to "do the obvious automatically, expect human help for the non-obvious". And as crazy as that sound, because any decent IDE would easily offer you to switch from interface to the single implementation class and back - this feature relies on full knowledge of the code base content - something that PHP simply does not have at runtime.
